sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -469,36 +589,129 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
+      private final ValueProvider<String> project;
+      @Nullable
+      private final ValueProvider<String> namespace;
+      @Nullable
+      private final String localhost;
+
+      private V1Options(ValueProvider<String> project, ValueProvider<String> namespace,
+          String localhost) {
+        this.project = project;
+        this.namespace = namespace;
+        this.localhost = localhost;
+      }
+
+      public static V1Options from(String projectId, String namespace, String localhost) {
+        return from(StaticValueProvider.of(projectId), StaticValueProvider.of(namespace),
+            localhost);
+      }
+
+      public static V1Options from(ValueProvider<String> project, ValueProvider<String> namespace,
+          String localhost) {
+        return new V1Options(project, namespace, localhost);
+      }
+
+      public String getProjectId() {
+        return project.get();
+      }
+
+      @Nullable
+      public String getNamespace() {
+        return namespace == null ? null : namespace.get();
+      }
+
+      public ValueProvider<String> getProjectValueProvider() {
+        return project;
+      }
+
+      @Nullable
+      public ValueProvider<String> getNamespaceValueProvider() {
+        return namespace;
+      }
+
+      @Nullable
+      public String getLocalhost() {
+        return localhost;
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder
+            .addIfNotNull(DisplayData.item("projectId", getProjectValueProvider())
+                .withLabel("ProjectId"))
+            .addIfNotNull(DisplayData.item("namespace", getNamespaceValueProvider())
+                .withLabel("Namespace"));
+      }
     }
 
     /**
-     * A class for v1 Cloud Datastore related options.
+     * A DoFn that translates a Cloud Datastore gql query string to {@code Query}.
      */
-    @VisibleForTesting
-    @AutoValue
-    abstract static class V1Options implements Serializable {
-      public static V1Options from(String projectId, Query query, @Nullable String namespace,
-                                   @Nullable String localhost) {
-        return new AutoValue_DatastoreV1_Read_V1Options(projectId, query, namespace, localhost);
+    static class GqlQueryTranslateFn extends DoFn<ValueProvider<String>, Query> {
+      private final V1Options v1Options;
+      private transient Datastore datastore;
+      private final V1DatastoreFactory datastoreFactory;
+
+      GqlQueryTranslateFn(V1Options options) {
+        this(options, new V1DatastoreFactory());
+      }
+
+      GqlQueryTranslateFn(V1Options options, V1DatastoreFactory datastoreFactory) {
+        this.v1Options = options;
+        this.datastoreFactory = datastoreFactory;
       }
 
-      public abstract String getProjectId();
-      public abstract Query getQuery();
-      @Nullable public abstract String getNamespace();
-      @Nullable public abstract String getLocalhost();
+      @StartBundle
```

Comment:
It's overkill to put this into StartBundle because the bundle will have only a single element.

It's also overkill to use a DoFn at all - even if there's a compelling argument against NestedValueProvider, at least, it should instead be a MapElements with a SerializableFunction.